### ISSUE TITLE
All inputs blocked during a break

### ISSUE THREAD

**Issue by Jehan** [codito/gnome-pomodoro#190]
I installed Pomodoro 2 days ago, and had it installed for a friend too. We both encountered the same issue.
When the break time occurred, usually we can still bypass it (though I haven't found a consistent way to do so. Sometimes just a wave with the mouse removes the dark screen, but sometimes it does not, but then Esc does). Well this one time, it just completely "froze" the desktop! It is not "frozen" as in "crashed". All software are still working fine and in motion (if there was music or video, you can still see it playing for instance). Simply we lost any capacity of interaction with the whole desktop and all the software in it! I even tried to reset GNOME-shell with alt-F2 + 'r', but even this was blocked (no command launcher appeared).
It seems like Pomodoro grabbed all inputs and refuses to release them. I tried to reset GNOME-shell from a TTY terminal instead, but unfortunately it actually crashed GNOME-shell. You can imagine this is bad, since both time it happened, we had work ongoing that got lost (luckily we often save so it was not the worst!).

P.S.: you could say that the concept of pomodoro is that we should not try to bypass the break, but well… sometimes I am just in the middle of something and stopping the work in the middle like this is worse than not having a break (since you lose the process, then waste time when you come back to it, then get tired and frustrated), whereas just postponing the break for 2 minutes later could end up in a much more relaxing break time. I still appreciate that the software reminds me to take a break soon (that I can just postpone in a few minutes, just the time to finish the work in progress). In any case the software should not control us!

**Comment by kamilprusko** [comment]
Made a minor change 791ea8e3a0ecafb25d80cde26e074b131f0e2390, but if the dialog is dismissed after the break it looks more like a problem with gnome-shell GrabHelper.

Have you tried waiting for the end of a break?

Are you using wayland? Could maybe other app request a grab at that time - popup menus, notifications from Chrome, Skype, Steam, etc., focused video app?

This error should be logged in

    journalctl /usr/bin/gnome-session

**Comment by Jehan** [comment]
> Have you tried waiting for the end of a break?

No, the 2 times it happened, I was like "fuck, my computer!" and I tried all I could to unblock it. Next time, if it happens again, I'll just wait.
For the time being, I have disabled the Screen Notifications and the Wake up screen (I believe one of these triggers the dark curtain with a timer over the desktop, and that the grab happens there, right?), but I will try to re-enable these (and just save my work all the time, which I already do anyway :p!).

> Are you using wayland?

Nope.

> Could maybe other app request a grab at that time - popup menus, notifications from Chrome, Skype, Steam, etc., focused video app?

I don't know. When it happened, in my case, I was coding in the terminal (I don't think it has any grab). In my friend's case, she was drawing in GIMP (there, we indeed have a few grabs in the code which may happen). Could there be some kind of race condition between pomodoro and another program? Is it what you are suggesting?

> This error should be logged in

So searching all errors mentioning pomodoro and/or a grab, I found these:

> Aug 24 14:15:00 darkmarmot gnome-session[1924]: Gjs-Message: JS LOG: Extension "pomodoro@arun.codito.in" had error: Extension point conflict: there is
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: Gjs-Message: JS LOG: Extension "pomodoro@arun.codito.in" had error: incorrect grab helper pop
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: Gjs-Message: JS LOG: Extension "pomodoro@arun.codito.in" had error: incorrect grab helper pop
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: (gnome-shell:2106): Gjs-WARNING **: JS ERROR: TypeError: grabHelper is undefined
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: _onCapturedEvent@resource:///org/gnome/shell/ui/grabHelper.js:17
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: (gnome-shell:2106): Gjs-WARNING **: JS ERROR: TypeError: grabHelper is undefined
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: _onCapturedEvent@resource:///org/gnome/shell/ui/grabHelper.js:17
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: (gnome-shell:2106): Gjs-WARNING **: JS ERROR: TypeError: grabHelper is undefined
> Aug 24 14:50:23 darkmarmot gnome-session[1924]: _onCapturedEvent@resource:///org/gnome/shell/ui/grabHelper.js:17
> Aug 24 14:50:27 darkmarmot gnome-session[1924]: (gnome-shell:2106): Gjs-WARNING **: JS ERROR: TypeError: grabHelper is undefined
> Aug 24 14:50:27 darkmarmot gnome-session[1924]: _onCapturedEvent@resource:///org/gnome/shell/ui/grabHelper.js:17
> Aug 24 14:50:27 darkmarmot gnome-session[1924]: (gnome-shell:2106): Gjs-WARNING **: JS ERROR: TypeError: grabHelper is undefined
> Aug 24 14:50:27 darkmarmot gnome-session[1924]: _onCapturedEvent@resource:///org/gnome/shell/ui/grabHelper.js:17

These errors actually go on for a while (counted 3266 lines), but they are basically all these same errors repeating over again and again (to make sure, I made a `sort|uniq` after removing the timestamp).
I can't remember for sure at what time it happened to me yesterday, but this looks like it may very well be the errors we are looking for. :-)

**Comment by kamilprusko** [comment]
I'm afraid that every pomodoro app will feel like it tries to control you, except for the apps that will ask for an action every time - then it may feel like you're babysitting these apps. I'm open to new ideas.

This app tries to work in the background for the most time and is a fairly forgiving. If you need to extend the break then just do it, disable the timer if you intend to use computer, the app will adjust to you.

**Comment by kamilprusko** [comment]
Great! Thanks for the logs.

Waiting for the end of break won't help here... :(

I'm not sure what causes the "invalid pop", but in grabHelper code:

    function _popGrabHelper(grabHelper) {
        let poppedHelper = _grabHelperStack.pop();
        if (poppedHelper != grabHelper)
            throw new Error("incorrect grab helper pop");

        if (_grabHelperStack.length == 0) {
            global.stage.disconnect(_capturedEventId);
            _capturedEventId = 0;
        }
    }

if "incorrect grab helper pop" occurs, event capturing may not get disabled.

I'll report it to gnome-shell bugzilla.

**Comment by Jehan** [comment]
> I'll report it to gnome-shell bugzilla.

Ok. Can you also give a link once it's reported? :-)

> I'm afraid that every pomodoro app will feel like it tries to control you

I feel you are pretty close to a good behavior already. But a little too violent still. I'm thinking this pomodoro app should be more like this slightly annoying but good friend telling you "hey dude, you are overworking! Come grab a beer".
But that should not be the "stupid" friend who thinks he helps you when he hides away your so important report which will make you miss a huge deal if you don't send it in 5 minutes.

> except for the apps that will ask for an action every time - then it may feel like you're babysitting these apps.

Actually I am not against having to click on my own to say "I start my break". The app could tell you "you should rest" but without being too intrusive (becoming very insistent only when you get really late on your break). Just be visible enough for you to remember "yes I should make a break" and as soon as you finish the current item, you can start your break.

I mean, I'm sure you know this too: you are coding on a complicated piece of code and are getting close to unravel the why and the how. And then… paf black screen! That's more likely to stress you out or frustrate you rather than prepare you for a calm break, right?
Ok so I am not a pomodoro expert or whatever (I discover this with this app) but I feel starting a break a few minutes earlier or later (and making a break a few minutes shorter or longer) is not a huge problem. Well I do hope exact timing is not the point of such a technique at least because strictness is more likely to get on someone's nerve than anything else. So such an app should be flexible too and adapt if you want to start earlier or later your breaks. Well… in my opinion.

> I'm open to new ideas.

I'll think a little and may propose some improvements later in a separate request. If I get a great idea, who knows? :-)

**Comment by kamilprusko** [comment]
https://bugzilla.gnome.org/show_bug.cgi?id=754102

**Comment by whatacold** [comment]
I've suffered from such a situation many times, often occurred after the break:
the keyboard doesn't work, but the mouse do.

Just a few minutes ago, i found a way to workaround this issue:
locking the screen in upper right corner using the mouse, and
the keyboard works again after unlocking.

HTH.

PS: my box is fedora 22, with Gnome.

**Comment by kamilprusko** [comment]
Thanks @whatacold for the tip.

**Comment by RichAyotte** [comment]
This bug gets me all the time. My workaround is to move my mouse to Activites, then <kbd>ALT</kbd>+<kbd>F2</kbd> to launch a command and then input <kbd>r</kbd> to restart the shell. Locking and unlocking doesn't always work for me.

```bash
$ gnome-shell --version
GNOME Shell 3.18.3
```
